File: elixir/lib/2022/07_no_space_left_on_device/aoc202207.ex
defmodule AOC2022.Day07 do
  @moduledoc """
  Advent of Code 2022, day 7: No Space Left On Device.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input),
    do: puzzle_input |> String.split("\n", trim: true) |> parse_files([""], %{"/" => 0})

  @doc """
  Parse shell commands to create a file tree.
  """
  def parse_files([], _, tree), do: tree

  def parse_files([<<"$ cd ", dir::binary>> | commands], cwd, tree),
    do: parse_files(commands, add_dir(cwd, dir), tree)

  def parse_files([<<"$ ls">> | commands], cwd, tree), do: parse_files(commands, cwd, tree)

  def parse_files([<<"dir ", dir::binary>> | commands], cwd, tree),
    do: parse_files(commands, cwd, Map.put(tree, repr_dir(cwd, dir), 0))

  def parse_files([file | commands], cwd, tree) do
    [size, name] = String.split(file)
    parse_files(commands, cwd, Map.put(tree, repr_dir(cwd, name), String.to_integer(size)))
  end

  defp add_dir([top | tree], dir) do
    case dir do
      "/" -> [""]
      ".." -> tree
      dir -> [dir, top | tree]
    end
  end

  defp repr_dir(dir, file) do
    [file | dir] |> Enum.reverse() |> Enum.join("/")
  end

  @doc """
  Solve part 1.
  """
  def part1(tree),
    do: tree |> dir_sizes() |> Enum.filter(fn size -> size <= 100_000 end) |> Enum.sum()

  @doc """
  Solve part 2.
  """
  def part2(tree) do
    target = (tree |> Map.values() |> Enum.sum()) - 40_000_000
    tree |> dir_sizes() |> Enum.sort() |> Enum.find(fn size -> size > target end)
  end

  @doc """
  Calculate the total size of each directory, including subdirectories.
  """
  def dir_sizes(tree) do
    Map.filter(tree, fn {_, size} -> size == 0 end)
    |> Map.keys()
    |> Enum.map(fn dir ->
      tree
      |> Map.filter(fn {name, _} -> String.starts_with?(name, dir <> "/") end)
      |> Enum.reduce(0, fn {_, size}, total -> total + size end)
    end)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2022/07_no_space_left_on_device/aoc202207.py
"""AoC 7, 2022: No Space Left On Device."""

# Standard library imports
import pathlib
import sys


def parse_data(puzzle_input):
    """Parse input."""
    tree = cwd = []
    parents = []
    for command in puzzle_input.split("\n"):
        match command.split():
            case ["$", "cd", "/"]:
                cwd = tree
                parents = []
            case ["$", "cd", ".."]:
                cwd = parents.pop()
            case ["$", "cd", _]:
                parents.append(cwd)
                cwd.append([])
                cwd = cwd[-1]
            case ["$", "ls"] | ["dir", _]:
                pass
            case [size, _]:
                cwd.append(int(size))
            case _:
                raise ValueError(f"{command!r} not parsed")
    return tree


def part1(tree):
    """Solve part 1."""
    return sum(size for size in dir_sizes(tree) if size <= 100_000)


def part2(tree):
    """Solve part 2."""
    sizes = dir_sizes(tree)
    target_size = max(sizes) - 40_000_000
    return min(size for size in sizes if size >= target_size)


def dir_sizes(tree):
    """Recursively calculate the size of each directory.

    ## Example:

    >>> dir_sizes([1, [2, 4, [8], 16], 32, [64, [128]], 256])
    [8, 30, 128, 192, 511]
    """
    dirs = []
    total_size = 0
    for content in tree:
        if isinstance(content, list):
            dirs.extend(dir_sizes(content))
            total_size += dirs[-1]
        else:
            total_size += content
    dirs.append(total_size)
    return dirs


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
